dhall/src/Dhall/Binary.hs

```diff
@@ -875,9 +876,14 @@ encodeExpressionInternal encodeEmbed = go
           where
             n32 = double2Float n64
 
+            n16 = toHalf n32
+
             useFloat = n64 == float2Double n32
 
-            useHalf = n64 == 0.0 || n64 == infinity || n64 == -infinity
+            -- Check if the original Double can be recovered from a half
+            halfEnough = n64 == (float2Double $ fromHalf n16)
+
+            useHalf = n64 == 0.0 || n64 == infinity || n64 == -infinity || halfEnough
```

Comment:
Can we get rid of those zero- and infinity checks too, now?

Reply:
Probably, according to `toHalf` source, it deals appropriately with infinity. Next change removes the special case.